scripts/with-go-mod.sh: use symlink instead of -modfile

thaJeztah opened the pull request:
While the "-modfile=vendor.mod" is slightly more correct, using a symlink allows for most of the go tools to work "as usual", just without an acual `go.mod` being committed in the repository.

**- Human readable description for the release notes**
<!--
Write a short (one line) summary that describes the changes in this
pull request for inclusion in the changelog.
It must be placed inside the below triple backticks section.

NOTE: Only fill this section if changes introduced in this PR are user-facing.
The PR must have a relevant impact/ label.
-->
```markdown changelog


```

**- A picture of a cute animal (not mandatory but encouraged)**

thaJeztah committed:
scripts/with-go-mod.sh: use symlink instead of -modfile

While the "-modfile=vendor.mod" is slightly more correct, using a
symlink allows for most of the go tools to work "as usual", just
without an acual `go.mod` being committed in the repository.

Signed-off-by: Sebastiaan van Stijn <github@gone.nl>

tianon commented on scripts/vendor:
```diff
@@ -31,7 +31,7 @@ outdated() {
     exit 1
   fi
-  (set -x ; go list -mod=vendor -mod=readonly -modfile=vendor.mod -u -m -json all | go-mod-outdated -update -direct)
+  (set -x ; go list -mod=vendor -mod=readonly -u -m -json all | go-mod-outdated -update -direct)
```
`go list -mod=vendor -mod=readonly` will just be `-mod=readonly` right? :thinking:

thaJeztah replied:
Good one; I recall `-mod=vendor` was needed to make it actually use vendor nowadays? Will need look them up 😂

thaJeztah replied:
Oh! yes, you're right `-mod` twice, I guess that's not combined, but overrides? (likely?)

tianon replied:
Yeah, it's like `-tags` which is one that feels like it _should_ combine but it overwrites (standard behavior of the stdlib `flag` library).

tianon commented on scripts/with-go-mod.sh:
```diff
@@ -6,28 +6,40 @@
-		EOF
-	} >&2
-else
+cleanup_paths=""
```
This should be an array, not a string:

```suggestion
cleanup_paths=()
```

tianon commented on scripts/with-go-mod.sh:
```diff
@@ -6,28 +6,40 @@
+	local target="$1"
+	local link="$2"
+
+	pushd "${ROOTDIR}" > /dev/null
+	trap 'popd > /dev/null' RETURN
```
Whoa whoa whoa, what?  This is wild, and unnecessary -- there are either flags on `ln` that can DTRT or `cp --symbolic-link`.  If you can't get the arguments finagled correctly, let me know and I'll happily help sort it out further.

thaJeztah replied:
❤️ I ran into the script not detecting that my own (relative) symlink was same as the symlink it wanted to create, so thought that changing to the directory would be the best way to "create a relative symlink here' and make it detect that my own symlink was "same" (so no warning printed) 🙈

tianon replied:
So maybe something like https://superuser.com/a/196655/101945? :eyes:

thaJeztah replied:
OK; I think I found why it originally didn't work; I used an absolute path for both sides; now using absolute path for only the "source" side, and that looks to work; creates relative symlink, and detects my own symlink for being "ok"

thaJeztah replied:
> https://superuser.com/a/196655/101945

Oh! GitHub didn't show your comment; that looks interesting as well yes; let me give that a try as well!

tianon replied:
If you use absolute path on both sides, you can use `ln --relative` to have `ln` automatically "fix" it for you.

thaJeztah replied:
Yeah; so ... macOS and BSD 🙈 - don't think they have that option 🫠

tianon replied:
!["I'm deleting them from my life." webcomic cover](https://github.com/user-attachments/assets/ebe10504-186c-4b2c-bd80-5d60455f9df5)

tianon commented on scripts/with-go-mod.sh:
```diff
@@ -6,28 +6,40 @@
+
 	set -x
+	ln -s "$target" "$link"
+	cleanup_paths="$cleanup_paths ${ROOTDIR}/$link"
```
```suggestion
	cleanup_paths+=( "${ROOTDIR}/$link" )
```

(frankly `"${ROOTDIR}/$link"` should go into a variable earlier in this function so it can just be used throughout)

tianon commented on scripts/with-go-mod.sh:
```diff
@@ -6,28 +6,40 @@
-	EOF
-	trap 'rm -f "${ROOTDIR}/go.mod"' EXIT
+if [ -n "$cleanup_paths" ]; then
+	trap 'rm -f'"$cleanup_paths" EXIT
```
```suggestion
	trap 'rm -f "${cleanup_paths[@]}"' EXIT
```

thaJeztah replied:
❤️ I never get the hang of these (array vars); will update

tianon commented on scripts/with-go-mod.sh:
```diff
@@ -6,28 +6,40 @@
-		go 1.23.0
-	EOF
-	trap 'rm -f "${ROOTDIR}/go.mod"' EXIT
+if [ -n "$cleanup_paths" ]; then
```
```suggestion
if [ "${#cleanup_paths[@]}" -gt 0 ]; then
```

thaJeztah committed:
scripts/with-go-mod.sh: simplify symlink comparison

Make sure the symlinks point to the same file; not if both are
identical (absolute or relative symlink); it's only for suppressing
the warning so not critical.

Signed-off-by: Sebastiaan van Stijn <github@gone.nl>
